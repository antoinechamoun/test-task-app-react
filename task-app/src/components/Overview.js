import { useEffect, useState } from "react";
import Form from "./Form";
import {
  addTaskFirebase,
  deleteTaskFirebase,
  loadTasks,
} from "../firebase/firebase";

const Overview = () => {
  const [tasks, setTasks] = useState([]);

  useEffect(() => {
    loadTasks().then((res) => setTasks(res));
  }, []);

  const handleValue = (e, id) => {
    const newTasks = tasks.map((task) => {
      if (task.id === id) {
        return { value: e.target.value, id: task.id, isEdit: true };
      }
      return task;
    });
    setTasks(newTasks);
  };

  const editTask = (id, isEdit) => {
    if (isEdit === false) {
      const newTasks = tasks.map((task) => {
        if (task.id === id) {
          return { value: task.value, id: task.id, isEdit: true };
        }
        return task;
      });
      setTasks(newTasks);
    } else {
      const newTasks = tasks.map((task) => {
        if (task.id === id) {
          return { value: task.value, id: task.id, isEdit: false };
        }
        return task;
      });
      setTasks(newTasks);
    }
  };

  const deleteTask = (id) => {
    const newTasks = tasks.filter((task) => task.id !== id);
    setTasks(newTasks);
  };
  return (
    <>
      <Form
        tasks={tasks}
        setTasks={setTasks}
        addTaskFirebase={addTaskFirebase}
      />
      <div data-testid="tasks-container">
        {tasks.length !== 0 &&
          tasks.map((task, id) => {
            const { value, isEdit } = task;
            return (
              <div key={id}>
                {isEdit ? (
                  <input
                    type="text"
                    value={value}
                    onChange={(e) => {
                      handleValue(e, task.id);
                    }}
                  />
                ) : (
                  <h3 data-testid="value-output">{value}</h3>
                )}
                <button onClick={() => editTask(task.id, isEdit)}>
                  {isEdit ? "Save" : "Edit"}
                </button>
                <button
                  onClick={() => {
                    deleteTask(task.id);
                    deleteTaskFirebase(task.id);
                  }}>
                  Delete
                </button>
              </div>
            );
          })}
      </div>
    </>
  );
};

export default Overview;
