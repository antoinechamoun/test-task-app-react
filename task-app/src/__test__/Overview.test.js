import { fireEvent, render, screen } from "@testing-library/react";
import Overview from "../components/Overview";
import userEvent from "@testing-library/user-event";
describe("Form Component", () => {
  it("renders an empty task initially", () => {
    const { container } = render(<Overview />);
    expect(container).toMatchSnapshot();
  });

  it("adds a new task after button click", () => {
    render(<Overview />);
    const input = screen.getByTestId("input-value");
    fireEvent.change(input, { target: { value: "Clean" } });
    const button = screen.getByRole("button", { name: "Add new task" });
    userEvent.click(button);

    expect(screen.getByTestId("value-output").textContent).toMatch(/Clean/i);
  });
});
