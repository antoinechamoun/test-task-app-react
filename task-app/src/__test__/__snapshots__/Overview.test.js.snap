// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component renders an empty task initially 1`] = `
<div>
  <form>
    <input
      data-testid="input-value"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Add new task
    </button>
  </form>
  <div
    data-testid="tasks-container"
  />
</div>
`;
